Add the shard's store status to the explain API

dakrone opened the pull request:
This adds information similar to what is from the [shard stores
API](https://www.elastic.co/guide/en/elasticsearch/reference/2.3/indices-shards-stores.html)
to the cluster allocation explanation API (in fact, internally it uses
that API).

This means when you have a decision that otherwise could indicate that a
shard can go somewhere, you now have more information:

```json
{
  "shard" : {
    "index" : "i",
    "index_uuid" : "QzoKda9aQCG_hCaZQ18GEg",
    "id" : 0,
    "primary" : true
  },
  "assigned" : false,
  "unassigned_info" : {
    "reason" : "CLUSTER_RECOVERED",
    "at" : "2016-04-11T20:58:04.088Z"
  },
  "allocation_delay" : "0s",
  "allocation_delay_ms" : 0,
  "remaining_delay" : "0s",
  "remaining_delay_ms" : 0,
  "nodes" : {
    "24Qmw4tdRTuVOtjAdtmr5Q" : {
      "node_name" : "Vampire by Night",
      "node_attributes" : { },
      "final_decision" : "YES",
      "weight" : 7.0,
      "decisions" : [ ],
      "store" : {
        "allocation_id" : "aC6qVWA7TT2pgsalYxxUJQ",
        "store_exception" : "IndexFormatTooOldException[Format version is not supported (resource BufferedChecksumIndexInput(SimpleFSIndexInput(path=\"/home/hinmanm/scratch/elasticsearch-5.0.0-alpha1-SNAPSHOT/data/elasticsearch/nodes/0/indices/QzoKda9aQCG_hCaZQ18GEg/0/index/segments_1\"))): -1906795950 (needs to be between 1071082519 and 1071082519). This version of Lucene only supports indexes created with release 5.0 and later.]",
        "allocation" : "UNUSED"
      }
    }
  }
}
```

The "store" section is the new section, and will include allocation, id,
and the exception if there is one.

Relates to #17372

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java:
```diff
@@ -76,18 +84,26 @@ public ClusterAllocationExplanation(StreamInput in) throws IOException {
         }
         this.nodeWeights = ntw;
         remainingDelayNanos = in.readVLong();
+        size = in.readVInt();
```
Maybe use readList and writeList?

dakrone replied:
Those require the object to be `Writeable`, which this is not

nik9000 replied:
It looks like readList doesn't but writeList does. There isn't a Streamable version. Oh well.

nik9000 replied:
Yikes. I guess it's been that long since I used it. Yes, the keyword.

nik9000 replied:
Meh, I don't really think it is worth it.

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java:
```diff
@@ -49,6 +54,9 @@
     private final long remainingDelayNanos;
+    private final List<IndicesShardStoresResponse.StoreStatus> shardStores;
+
+    private Map<DiscoveryNode, IndicesShardStoresResponse.StoreStatus> nodeStoreStatus = null;
```
I wonder if we should use `@Transient` for this. I mean, we don't use java's serialization but it is kind of a useful signal.

dakrone replied:
Do you mean the `transient` keyword? I'm not familiar with the `@Transient` annotation at all

bleskes commented:
This is an awesome extension. I'm a bit worried by the output, if you don't know the internals - this sounds like we have decided to allocation the shard (due to the final decision), but we won't because the store is corrupted (unless this is a replica):

```
"final_decision" : "YES",
      "weight" : 7.0,
      "decisions" : [ ],
      "store" : {
        "allocation_id" : "aC6qVWA7TT2pgsalYxxUJQ",
        "store_exception" : "IndexFormatTooOldException[Format version is not supported (resource BufferedChecksumIndexInput(SimpleFSIndexInput(path=\"/home/hinmanm/scratch/elasticsearch-5.0.0-alpha1-SNAPSHOT/data/elasticsearch/nodes/0/indices/QzoKda9aQCG_hCaZQ18GEg/0/index/segments_1\"))): -1906795950 (needs to be between 1071082519 and 1071082519). This version of Lucene only supports indexes created with release 5.0 and later.]",
        "allocation" : "UNUSED"
      }
```

Also, I'm not sure the allocation section of the shard stores is relevant here.

dakrone commented:
@bleskes with the output though, I didn't want to conflate the two (deciders and store) in the output because they aren't conflated in ES (completely separate things).

> Also, I'm not sure the allocation section of the shard stores is relevant here. 

I can take it out if desired, but I thought it would be good to show as much info as possible, it can at least show the allocation status of the store if there are multiple unused allocations (makes it more clear that the store is in a bad state if it is marked as unused and has an exception)

clintongormley commented:
> this sounds like we have decided to allocation the shard (due to the final decision), but we won't because the store is corrupted (unless this is a replica):

you're right.  it does sound like that.  what does it actually mean?

ywelsch commented:
>  I didn't want to conflate the two (deciders and store) in the output because they aren't conflated in ES (completely separate things).

I feel that just putting shard stores information here is not sufficient. The explain API needs to explain why a shard is unallocated (e.g. no shard data found or no shard data with matching allocation id found, i.e. only stale copies). The good thing about allocation deciders is that they explicitly state what the issue is.

Maybe the explanation API can abstract from the implementation details of allocation deciders and just show how the steps that lead to an allocation decision to assign / not assign a shard. Such a step can be a decision from an allocation decider as well as other checks.

bleskes commented:
>  what does it actually mean?

Not sure exactly, which part qualifies as "it" here, so I hope I clarify it but saying that the allocation deciders will not block the assignment on that node but a primary can not be assigned to it because the data copy on it is not valid.

> Maybe the explanation API can abstract from the implementation details

It feels like this is indeed required in order to come up with an output that makes sense from this API's perspective alone. An alternative route it to have the decision made by the primary shard allocator wrapped in a allocation decider format.

dakrone commented:
Thanks for taking a look @bleskes, @ywelsch, and @clintongormley - I'll try to re-work the output to be friendlier for people not familiar with the concept of a store but still keep the good debugging information. I think I may use the `final_decision` as well for this (perhaps with an "ERROR" option when the store is corrupted)

dakrone commented:
I changed the output for this -

Happy case (meaning no corruption) output:

``` json
{
  "shard" : {
    "index" : "i",
    "index_uuid" : "P4WksCZdS2ukiiNa6eF1Gw",
    "id" : 0,
    "primary" : false
  },
  "assigned" : false,
  "unassigned_info" : {
    "reason" : "CLUSTER_RECOVERED",
    "at" : "2016-04-14T19:55:02.019Z"
  },
  "allocation_delay" : "0s",
  "allocation_delay_ms" : 0,
  "remaining_delay" : "0s",
  "remaining_delay_ms" : 0,
  "nodes" : {
    "bG9sYnmJQiGuBz-dy_PJ2A" : {
      "node_name" : "Andrew Chord",
      "node_attributes" : { },
      "store" : {
        "shard_copy" : "PRIMARY"
      },
      "final_decision" : "NO",
      "weight" : 5.0,
      "decisions" : [ {
        "decider" : "same_shard",
        "decision" : "NO",
        "explanation" : "the shard cannot be allocated on the same node id [bG9sYnmJQiGuBz-dy_PJ2A] on which it already exists"
      } ]
    }
  }
}
```

This now only indicates the copy of the data that the node is using (in the
`shard_copy`), either primary or replica (or unused if there is an extra copy of
the data around that has not been removed).

Sad case output:

``` json
{
  "shard" : {
    "index" : "i",
    "index_uuid" : "P4WksCZdS2ukiiNa6eF1Gw",
    "id" : 0,
    "primary" : true
  },
  "assigned" : false,
  "unassigned_info" : {
    "reason" : "CLUSTER_RECOVERED",
    "at" : "2016-04-14T19:57:25.753Z"
  },
  "allocation_delay" : "0s",
  "allocation_delay_ms" : 0,
  "remaining_delay" : "0s",
  "remaining_delay_ms" : 0,
  "nodes" : {
    "SWw_UzplSx-OlkshdZBw5A" : {
      "node_name" : "Letha",
      "node_attributes" : { },
      "store" : {
        "store_exception" : "CorruptIndexException[misplaced codec footer (file extended?): remaining=36, expected=16, fp=303 (resource=BufferedChecksumIndexInput(SimpleFSIndexInput(path=\"/home/hinmanm/scratch/elasticsearch-5.0.0-alpha1-SNAPSHOT/data/elasticsearch/nodes/0/indices/P4WksCZdS2ukiiNa6eF1Gw/0/index/segments_2\")))]",
        "shard_copy" : "UNUSED"
      },
      "final_decision" : "STORE_ERROR",
      "weight" : 6.0,
      "decisions" : [ ]
    }
  }
}
```

The exception is still show, as well as the shard copy being "UNUSED" (because
it's corrupt), but now the `final_decision` has changed to `STORE_ERROR` because
the shard store is corrupt and can't be used at all.

Allocation id and legacy id are gone from both outputs (these aren't really
useful for humans to know)

clintongormley commented:
Better, easier to understand.

> The exception is still show, as well as the shard copy being "UNUSED" (because it's corrupt), but now the final_decision has changed to STORE_ERROR because the shard store is corrupt and can't be used at all.

I assume this means that there is no valid copy of the shard available on any node?  Wondering if `STORE_ERROR` should be `NO_PRIMARY_AVAILABLE` or something like that?

dakrone commented:
> I assume this means that there is no valid copy of the shard available on any node?

`final_decision` is still per-node, so only on that node is there a STORE_ERROR

> Wondering if STORE_ERROR should be NO_PRIMARY_AVAILABLE or something like that?

I don't think this makes sense, since not having a primary and having a primary but it's got an error are two different things

bleskes commented:
I think the new version is clearer as well. Thx Lee. Double checking - what happens when the shard we're explaining is a replica? In this case we may still allocated it to that node (but prefer others good copies to avoid data migration).

> Wondering if STORE_ERROR should be NO_PRIMARY_AVAILABLE or something like that?

I would prefer STORE_ERROR as well. This is a per-node output and NO_PRIMARY_AVAILABLE sounds like a global cluster wide things (no node has a valid copy).

Last, when designing the output, I think it will be useful to consider what Yannick already mentioned - it may be that we found valid good stores, the allocation deciders are all good to go and still we don't allocated a shard because none of the allocation IDs match the previously active allocation ids in the meta data (i.e., the found copies are stale). How are we planning to reflect that? it might be as simple as saying `STALE_COPY` ?

dakrone commented:
> what happens when the shard we're explaining is a replica? In this case we may still allocated it to that node (but prefer others good copies to avoid data migration).

The first example I posted was explaining a replica, the `shard_copy` is `PRIMARY` because that node currently has the primary copy of that shard, if other nodes had a replica copy it would have `"shard_copy": "REPLICA"` if it could be used or `"shard_copy": "UNUSED"` if there was a reason it couldn't be used (corruption). If there are no copies of the data, it'll just be `"store": {}`.

> it may be that we found valid good stores, the allocation deciders are all good to go and still we don't allocated a shard because none of the allocation IDs match the previously active allocation ids in the meta data (i.e., the found copies are stale). How are we planning to reflect that? it might be as simple as saying `STALE_COPY` ?

Yes, I added a commit to track stale allocation ids and it will report `STALE_COPY` if a stale copy is found

dakrone commented:
@ywelsch since @bleskes is on vacation, could you review this? I'd like to get it in for the 5.0 release

ywelsch commented:
Had a first go at this. Two observations I would like to clarify first before diving deeper:

I think that using the allocation status of the store status `storeStatus.getAllocationStatus()` does not add anything to the output. This information repeats what is already captured by `shard.primary` and `assigned`. In addition it can be inconsistent with `shard.primary` and `assigned` as it might be based on a different cluster state (the one that's the most recent on the node from which we fetch the shard store information which can be a different than the one on the coordinating node), making it quite confusing.

In my opinion store information should convey exactly the following information:
1. Whether a shard copy is available or not
2. Whether it is a corrupt copy
3. Whether it is a stale copy

For uniformity I would prefer that the `shard_copy` field captures all these cases:
- For 1) I would prefer we explicitly mark `shard_copy` as `NONE` if no data is available on disk.
- For 2) I would like us to distinguish whether the copy is corrupt (checksums mismatch) or we cannot read the disk properly (Hopefully there is a way to distinguish this based on exception type). Then mark `shard_copy` as `CORRUPT` or `IO_ERROR` depending on the situation and add the `store_exception` (as you did).
- For 3) I would mark `shard_copy` as `STALE`.
- In case data is available and neither corrupt nor stale I would mark "shard_copy" as AVAILABLE.

The final decision should then simply be "YES", "NO", or "ALREADY_ASSIGNED".
In case of "NO" it should also have a human readable explanation (e.g. "shard copy is stale")

Second observation:
ClusterAllocationExplanation also has too much logic in `toXContent`. By using the Java API I would like to get the same information. This means that there should be an object reflecting the store information with a getter to access it.

dakrone commented:
Hi @ywelsch, thanks for the great feedback! I totally agree!

I've changed the `ClusterAllocationExplanation` to have no logic in `toXContent` and make everything available via the Java API, I've also added the final decision and explanation as well as the store copy output, here's what it looks like now:

Happy Case (non-corrupt):

```json
{
  "shard" : {
    "index" : "i",
    "index_uuid" : "uPnPR2pPSHOCU5bG-KVB-g",
    "id" : 0,
    "primary" : false
  },
  "assigned" : false,
  "unassigned_info" : {
    "reason" : "INDEX_CREATED",
    "at" : "2016-04-25T18:26:42.946Z"
  },
  "allocation_delay" : "0s",
  "allocation_delay_ms" : 0,
  "remaining_delay" : "0s",
  "remaining_delay_ms" : 0,
  "nodes" : {
    "Ocyk1ql4QaejNBP6UvMZhA" : {
      "node_name" : "Vegas",
      "node_attributes" : {
        "testattr" : "test"
      },
      "store" : {
        "shard_copy" : "AVAILABLE"
      },
      "final_decision" : "NO",
      "final_explanation" : "the shard cannot be assigned because one or more allocation decider returns a 'NO' decision",
      "weight" : 5.0,
      "decisions" : [ {
        "decider" : "same_shard",
        "decision" : "NO",
        "explanation" : "the shard cannot be allocated on the same node id [Ocyk1ql4QaejNBP6UvMZhA] on which it already exists"
      } ]
    }
  }
}
```

Corrupt case:

```json
{
  "shard" : {
    "index" : "i",
    "index_uuid" : "WxAuxGpdSX6ezspqgtsMyQ",
    "id" : 0,
    "primary" : true
  },
  "assigned" : false,
  "unassigned_info" : {
    "reason" : "CLUSTER_RECOVERED",
    "at" : "2016-04-25T18:36:03.208Z"
  },
  "allocation_delay" : "0s",
  "allocation_delay_ms" : 0,
  "remaining_delay" : "0s",
  "remaining_delay_ms" : 0,
  "nodes" : {
    "uyixwqV1Rfuwmot-gDL8IQ" : {
      "node_name" : "Shiver Man",
      "node_attributes" : { },
      "store" : {
        "shard_copy" : "CORRUPT",
        "store_exception" : "CorruptIndexException[misplaced codec footer (file extended?): remaining=32, expected=16, fp=179 (resource=BufferedChecksumIndexInput(SimpleFSIndexInput(path=\"/home/hinmanm/scratch/elasticsearch-5.0.0-SNAPSHOT/data/elasticsearch/nodes/0/indices/WxAuxGpdSX6ezspqgtsMyQ/0/index/segments_2\")))]"
      },
      "final_decision" : "NO",
      "final_explanation" : "the copy of data in the shard store is corrupt",
      "weight" : 6.0,
      "decisions" : [ ]
    }
  }
}
```

Please take another look and let me know what you think!

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java:
```diff
@@ -45,49 +53,57 @@
+        final Map<DiscoveryNode, Decision> nodeDecisions = nodeToDecision == null ? Collections.emptyMap() : nodeToDecision;
+        final Map<DiscoveryNode, Float> nodeWeights = nodeToWeight == null ? Collections.emptyMap() : nodeToWeight;
+        assert nodeDecisions.size() == nodeWeights.size() : "decision and weight list should be the same size";
+        final Map<DiscoveryNode, IndicesShardStoresResponse.StoreStatus> storeStatuses = calculateStoreStatuses(shardStores);
```
I like that you separated the stats calculation from toXContent. I want to go one step further though (my previous comment was not explicit enough in that regard): Similar to other code (e.g. `NodeStats`), I want `ClusterAllocationExplanation` to just be a "result" object. This means that all the computation should be happening in `TransportClusterAllocationExplainAction`. The explanation does not need `activeAllocationIds`.

dakrone replied:
Makes sense, I will move everything there

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java:
```diff
@@ -96,27 +112,85 @@ public void writeTo(StreamOutput out) throws IOException {
+
+        if (node.getId().equals(assignedNodeId)) {
+            finalDecision = FinalDecision.ALREADY_ASSIGNED;
+            finalExplanation = "the shard is already assigned to this node";
```
The example you gave in your last comment shows only single allocation decider same_shard as "NO" but "final_decision" is "NO" instead of "ALREADY_ASSIGNED". Is there an error in the example?

dakrone replied:
Yeah, I fixed this and added some tests :)

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java:
```diff
@@ -96,27 +112,85 @@ public void writeTo(StreamOutput out) throws IOException {
+        if (storeStatus != null) {
+            final Throwable storeErr = storeStatus.getStoreException();
+            if (storeErr != null) {
+                finalDecision = FinalDecision.NO;
```
The store has no impact on the decision whether a replica shard can be allocated to a node. This means that if the store for the existing data is corrupt we will still allocate a replica to that node (and wipe the folder). For a primary, it only plays a role if we require data to be there (`ShardRouting.allocatedPostIndexCreate() == false`) and the data is STALE.

dakrone replied:
Updated this with your notes

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java:
```diff
@@ -96,27 +112,85 @@ public void writeTo(StreamOutput out) throws IOException {
+                    finalExplanation = "there was an IO error reading from data in the shard store";
+                }
+            } else if (activeAllocationIds.isEmpty() || activeAllocationIds.contains(storeStatus.getAllocationId())) {
+                // If either we don't have allocation IDs, or they contain the store allocation id, show the allocation
```
this comment is outdated

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java:
```diff
@@ -180,30 +251,220 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
-                builder.endObject(); // end node <uuid>
+            for (NodeExplanation explanation : nodeExplanations.values()) {
+                explanation.toXContent(builder, params);
             }
```
I checked out the code and saw the lines

```
builder.field("allocation_delay", TimeValue.timeValueNanos(delay));
builder.field("allocation_delay_ms", TimeValue.timeValueNanos(delay).millis());
builder.field("remaining_delay", TimeValue.timeValueNanos(remainingDelayNanos));
builder.field("remaining_delay_ms", TimeValue.timeValueNanos(remainingDelayNanos).millis());
```

There is a method `XContentBuilder.timeValueField` that automatically does "the right thing" by displaying the human readable time value when `?h` is specified.

dakrone replied:
Definitely better, added that.

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java:
```diff
@@ -138,22 +212,19 @@ public UnassignedInfo getUnassignedInfo() {
-    /** Return a map of node to decision for shard allocation */
-    public Map<DiscoveryNode, Decision> getNodeDecisions() {
-        return this.nodeToDecision;
+    /** Return the remaining allocation delay for this shard in nanoseconds */
```
as rest api displays this as millis, we should also return millis here. (Best would be to directly construct ClusterAllocationExplanation with remainingDelayMillis.)

dakrone replied:
Sure, I changed this to millis

ywelsch commented:
@dakrone I left more comments. The trickiest part is probably how store information influences (or does not influence) final decision. I also wonder in what kind of situations we can run into `activeAllocationIds.empty() == true` and whether we can provide a more helpful message in that case.

dakrone commented:
Thanks again @ywelsch, I pushed more commits with your feedback.

> I also wonder in what kind of situations we can run into activeAllocationIds.empty() == true and whether we can provide a more helpful message in that case.

So this occurs when recovering an older index that doesn't have allocation ids yet, I changed this to have a separate `StoreCopy` status and added a TODO to get the store info the "old way" in the future. This PR is already getting big so I'll leave it for the next one :)

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java:
```diff
@@ -138,22 +114,14 @@ public UnassignedInfo getUnassignedInfo() {
-     */
-    public Map<DiscoveryNode, Float> getNodeWeights() {
-        return this.nodeWeights;
+    /** Return the remaining allocation delay for this shard in nanoseconds */
```
milliseconds

dakrone replied:
Whoops, fixed

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplanation.java:
```diff
@@ -174,36 +142,118 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
-                builder.field("remaining_delay", TimeValue.timeValueNanos(remainingDelayNanos));
-                builder.field("remaining_delay_ms", TimeValue.timeValueNanos(remainingDelayNanos).millis());
+                builder.timeValueField("allocation_delay_ms", "allocation_delay", TimeValue.timeValueNanos(delay));
+                builder.timeValueField("remaining_delay_ms", "remaining_delay", TimeValue.timeValueMillis(remainingDelayMillis));
```
In most places (unfortunately not all) we use `_in_millis` as a suffix for the human readable form.

dakrone replied:
Okay, I've changed this to `_in_millis`

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportClusterAllocationExplainAction.java:
```diff
@@ -119,11 +130,74 @@ public static Decision tryShardOnNode(ShardRouting shard, RoutingNode node, Rout
+        } else {
+            // No copies of the data, so deciders are what influence the decision and explanation
+            storeCopy = ClusterAllocationExplanation.StoreCopy.NONE;
+        }
```
I think this method is not quite right yet. A few observations:
- if (unassigned) primary, then finalExplanation / finalDecision should be influenced by staleness of copy. In case of a replica, however, staleness does not influence allocation decision.
- if replica, then we can still show that store copy is corrupt / has an io / error without influencing final decision / explanation
- for the primary / replica shard allocator, corrupt / io_error is treated as no data.

I think we can first calculate store copy (which just represents the status on disk) and then influence finalExpl / finalDecision based on that.

Here is my go at it:

```
    public static NodeExplanation calculateNodeExplanation(ShardRouting shard,
                                                           DiscoveryNode node,
                                                           Decision nodeDecision,
                                                           Float nodeWeight,
                                                           IndicesShardStoresResponse.StoreStatus storeStatus,
                                                           String assignedNodeId,
                                                           Set<String> activeAllocationIds) {

        final StoreCopy storeCopy;

        if (storeStatus == null) {
            // No copies of the data
            storeCopy = StoreCopy.NONE;
        } else {
            final Throwable storeErr = storeStatus.getStoreException();
            if (storeErr != null) {
                if (ExceptionsHelper.unwrapCause(storeErr) instanceof CorruptIndexException) {
                    storeCopy = StoreCopy.CORRUPT;
                } else {
                    storeCopy = StoreCopy.IO_ERROR;
                }
            } else if (activeAllocationIds.isEmpty()) {
                // The ids are only empty if dealing with a legacy index
                // TODO: fetch the shard state versions and display here?
                storeCopy = StoreCopy.UNKNOWN;
            } else if (activeAllocationIds.contains(storeStatus.getAllocationId())) {
                storeCopy = StoreCopy.AVAILABLE;
            } else {
                // Otherwise, this is a stale copy of the data (allocation ids don't match)
                storeCopy = StoreCopy.STALE;
            }
        }

        final FinalDecision finalDecision;
        final String finalExplanation;

        if (node.getId().equals(assignedNodeId)) {
            finalDecision = FinalDecision.ALREADY_ASSIGNED;
            finalExplanation = "the shard is already assigned to this node";
        } else if (shard.primary() && shard.unassigned() && storeCopy == StoreCopy.STALE) {
            finalExplanation = "the copy of the shard is stale, allocation ids do not match";
            finalDecision = FinalDecision.NO;
        } else {
            if (nodeDecision.type() == Decision.Type.NO) {
                finalDecision = FinalDecision.NO;
                finalExplanation = "the shard cannot be assigned because one or more allocation decider returns a 'NO' decision";
            } else {
                finalDecision = FinalDecision.YES;
                if (storeCopy == StoreCopy.AVAILABLE) {
                    finalExplanation = "the shard can be assigned and the node contains a valid copy of the shard data";
                } else {
                    finalExplanation = "the shard can be assigned";
                }
            }
        }

        return new NodeExplanation(node, nodeDecision, nodeWeight, storeStatus, finalDecision, finalExplanation, storeCopy);
    }
```

dakrone replied:
Thanks! I'll use this :)

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportClusterAllocationExplainAction.java:
```diff
@@ -139,14 +213,35 @@ public static ClusterAllocationExplanation explainShard(ShardRouting shard, Rout
-            final Settings indexSettings = metadata.index(shard.index()).getSettings();
-            remainingDelayNanos = ui.getRemainingDelay(System.nanoTime(), metadata.settings(), indexSettings);
+            final Settings indexSettings = indexMetaData.getSettings();
+            long remainingDelayNanos = ui.getRemainingDelay(System.nanoTime(), metadata.settings(), indexSettings);
```
just noticed this uses `metadata.settings()` (= correct) whereas `updateDelay` in `AllocationService` uses node Settings (which are not updated).

ywelsch replied:
Looking at this again, I think we can remove the node settings as updateDelay / getRemainingDelay only depends on index settings.

dakrone replied:
Okay, I can remove that, but maybe in a different PR since it's a little outside the scope for this one?

ywelsch replied:
sounds good.

ywelsch commented:
I like the new structure! I added more comments about `calculateNodeExplanation`, the method that encapsulates now the essential logic of this PR. Can you add a few more unit tests for that method? I also wonder whether we should separate the testing of JSON serialization and generation of explanation result (now that these classes are separated). For example, `testStaleShardExplanation` does a little bit of both.

dakrone commented:
@ywelsch okay, I added more tests for the node explanation calculation. For the JSON serialization, I test the serialization of the both the CAE and NE in there, I'm not sure if needs further explanation? (there is also an integ test that tests serialization).

Let me know what you think, I think it's getting close!

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/cluster/allocation/TransportClusterAllocationExplainAction.java:
```diff
@@ -141,51 +141,57 @@ public static NodeExplanation calculateNodeExplanation(ShardRouting shard,
+        if (node.getId().equals(assignedNodeId)) {
+            finalDecision = ClusterAllocationExplanation.FinalDecision.ALREADY_ASSIGNED;
+            finalExplanation = "the shard is already assigned to this node";
+        } else if (shard.primary() && shard.unassigned() && storeCopy == ClusterAllocationExplanation.StoreCopy.STALE) {
```
My (hopefully) last corrections here (in this case I'm correcting my earlier self 😞): 
- We should also add allocatedPostIndexCreate() to be more in line with PrimaryShardAllocator. This means that this condition and the subsequent two become `shard.primary() && shard.unassigned() && shard.allocatedPostIndexCreate(indexMetaData) && storeCopy == ClusterAllocationExplanation.StoreCopy.STALE`

- We should also handle `StoreCopy.NONE` in the same way. In case of assigning an unassigned primary, having no shard data is also a NO.

dakrone replied:
Right you are! I've added this and tests for this specific case

ywelsch commented:
@dakrone left two more minor comments regarding node explanation calculation, otherwise this LGTM. Thanks for adding this feature, it's much appreciated!

clintongormley commented:
@dakrone looks like the docs still need writing?

dakrone commented:
@clintongormley yep, I'm going to open a separate PR for the docs
